stacks-signer/src/v0/signer.rs

```diff
@@ -0,0 +1,318 @@
+    /// The signer id
+    pub signer_id: u32,
+    /// The signer slot ids for the signers in the reward cycle
+    pub signer_slot_ids: Vec<SignerSlotID>,
+    /// The addresses of other signers
+    pub signer_addresses: Vec<StacksAddress>,
+    /// The reward cycle this signer belongs to
+    pub reward_cycle: u64,
+    /// SignerDB for state management
+    pub signer_db: SignerDb,
+}
+
+impl std::fmt::Display for Signer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Cycle #{} Signer #{}", self.reward_cycle, self.signer_id,)
+    }
+}
+
+impl SignerTrait<SignerMessage> for Signer {
+    /// Create a new signer from the given configuration
+    fn new(config: SignerConfig) -> Self {
+        Self::from(config)
+    }
+
+    /// Refresh the next signer data from the given configuration data
+    fn update_signer(&mut self, _new_signer_config: &SignerConfig) {
+        // do nothing
+    }
+    /// Return the reward cycle of the signer
+    fn reward_cycle(&self) -> u64 {
+        self.reward_cycle
+    }
+
+    /// Process the event
+    fn process_event(
+        &mut self,
+        stacks_client: &StacksClient,
+        event: Option<&SignerEvent<SignerMessage>>,
+        _res: Sender<Vec<OperationResult>>,
+        current_reward_cycle: u64,
+    ) {
+        let event_parity = match event {
+            Some(SignerEvent::BlockValidationResponse(_)) => Some(current_reward_cycle % 2),
+            // Block proposal events do have reward cycles, but each proposal has its own cycle,
+            //  and the vec could be heterogenous, so, don't differentiate.
+            Some(SignerEvent::MinerMessages(..))
+            | Some(SignerEvent::NewBurnBlock(_))
+            | Some(SignerEvent::StatusCheck)
+            | None => None,
+            Some(SignerEvent::SignerMessages(msg_parity, ..)) => Some(u64::from(*msg_parity) % 2),
+        };
+        let other_signer_parity = (self.reward_cycle + 1) % 2;
+        if event_parity == Some(other_signer_parity) {
+            return;
+        }
+        debug!("{self}: Processing event: {event:?}");
+        match event {
+            Some(SignerEvent::BlockValidationResponse(block_validate_response)) => {
+                debug!("{self}: Received a block proposal result from the stacks node...");
+                self.handle_block_validate_response(block_validate_response)
+            }
+            Some(SignerEvent::SignerMessages(_signer_set, messages)) => {
+                debug!(
+                    "{self}: Received {} messages from the other signers. Ignoring...",
+                    messages.len()
+                );
+            }
+            Some(SignerEvent::MinerMessages(messages, _)) => {
+                if current_reward_cycle != self.reward_cycle {
+                    // There is not point in processing blocks if we are not the current reward cycle (we can never actually contribute to signing these blocks)
```

Comment:
I don't think this is true. If you're in the prepare phase for cycle N+1 (i.e. you're at the end of cycle N), you already have the reward set and thus signing keys for cycle N+1's signers, so you ought to be able to sign for blocks in cycle N+1.

Reply:
Yes you could sign for cycle N+1. But this check is specifically to check "are we in cycle N+1"? We have a signer running for N and N+1. If we are in N, N+1 signers should not sign anything (we can be in its prepare phase, but they are not the approved signing set for blocks in reward cycle N. so should sign nothing). If we are in cycle N+1, but we are a cycle N signer, the same applies and we should not sign the block. hence this check.